lib/devices.go

```diff
@@ -0,0 +1,238 @@
+package usb_tree_lib
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/google/gousb"
+	"github.com/google/gousb/usbid"
+)
+
+const (
+	StateNormal  LogState = "normal"
+	StateAdded   LogState = "added"
+	StateRemoved LogState = "removed"
+)
+
+type LogState string
+
+type Device struct {
+	Path      []int
+	Name      string
+	VendorId  string
+	ProductId string
+	Speed     string
+	Bus       int
+	State     LogState
+}
+
+type TreeNode struct {
+	Device
+	Children []TreeNode
+}
+
+type Log struct {
+	Time  time.Time
+	Text  string
+	State LogState
+}
+
+var (
+	cachedDevices []Device
+	lastMergedMap map[string]Device
+	logs          []Log
+)
+
+var pollingStop chan bool
+
+func Stop() {
+	select {
+	case pollingStop <- true:
+	default:
+	}
+}
+
+func Init(onUpdateCallback func([]Device)) []Device {
+	pollingStop = make(chan bool)
+	Refresh()
+	go func() {
+		ticker := time.NewTicker(250 * time.Millisecond)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				newDevices := getDevices()
+				changed, mergedDevices := deviceDiff(newDevices)
+				if changed {
+					onUpdateCallback(mergedDevices)
+				}
+			case <-pollingStop:
+				close(pollingStop)
+				return
+
```

Comment:
This empty line probably shouldn't exist, but I feel one should exist between lines 69 and 70. What do you think?